src/constraint/encoding.ts

```diff
@@ -389,10 +385,10 @@ export const ENCODING_CONSTRAINTS: EncodingConstraintModel[] = [
       return true;
     }
   }
-].map((ec: EncodingConstraint) => new EncodingConstraintModel(ec));
+].map((ec: EncodingConstraint<FieldQuery>) => new EncodingConstraintModel<FieldQuery>(ec));
 
-export const ENCODING_CONSTRAINT_INDEX: {[name: string]: EncodingConstraintModel} =
-  ENCODING_CONSTRAINTS.reduce((m, ec: EncodingConstraintModel) => {
+export const ENCODING_CONSTRAINT_INDEX: {[name: string]: EncodingConstraintModel<FieldQuery>} =
```

Comment:
FIELD_QUERY_CONSTRAINT_INDEX?   (and then you implement a separate collection/index for ValueQuery.)